src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -1271,12 +1266,25 @@ void kds_fini_client(struct kds_sched *kds, struct kds_client *client)
 	struct kds_client_hw_ctx *curr = NULL;
 	struct kds_client_ctx *c_curr = NULL;
 
-	/* Release legacy client's resources */
-	_kds_fini_client(kds, client, client->ctx);
+	/* Release legacy client's resources for XOCL */
+	if (client->ctx) {
+		_kds_fini_client(kds, client, client->ctx);
+		mutex_lock(&client->lock);
+		curr = kds_get_hw_ctx_by_id(client, 0 /* default hw cx id */);
+		if (curr)
+			kds_free_hw_ctx(client, curr);
+		mutex_unlock(&client->lock);
+	}
+	/* this is for ZOCL legacy client */
 	if(!list_empty(&client->ctx_list)) {
 		list_for_each_entry(c_curr, &client->ctx_list, link) {
 			_kds_fini_client(kds, client, c_curr);
 		}
+		mutex_lock(&client->lock);
+		curr = kds_get_hw_ctx_by_id(client, 0 /* default hw cx id */);
+		if (curr)
+			kds_free_hw_ctx(client, curr);
+		mutex_unlock(&client->lock);
```

Comment:
The xocl and zocl cases are similar. These 5 lines of code are exactly the same. Can we unify? This is the point of kds_core.c.

Reply:
Fixed. Please check